File: runinator-command-center/src/utils/url-sync.ts
// pure mapping between the URL hash and (tab, id) route state, split out from useUrlSync so it can
// be unit-tested without a DOM. the hash layout is #/<Tab> or #/<Tab>/<id>.

export interface ParsedRoute {
  tab: string | null;
  id: string | null;
}

export function parseRoute(hash: string, isKnownTab: (tab: string) => boolean): ParsedRoute {
  const raw = hash.replace(/^#\/?/, "");

  if (!raw) {
    return { tab: null, id: null };
  }

  const [tabPart, idPart] = raw.split("/");
  const tab = isKnownTab(tabPart) ? tabPart : null;
  return { tab, id: idPart ? decodeURIComponent(idPart) : null };
}

export function formatRoute(tab: string, id: string | null): string {
  return id ? `#/${tab}/${encodeURIComponent(id)}` : `#/${tab}`;
}

File: runinator-command-center-flutter/lib/core/utils/url_sync.dart
// port of core/utils/url-sync.ts.
//
// pure mapping between a URL hash and (tab, id) route state, kept dependency-free
// so it can be unit-tested without a browser. the hash layout is #/<Tab> or
// #/<Tab>/<id>.

class ParsedRoute {
  const ParsedRoute({required this.tab, required this.id});

  final String? tab;
  final String? id;
}

ParsedRoute parseRoute(String hash, bool Function(String tab) isKnownTab) {
  final raw = hash.replaceFirst(RegExp(r'^#/?'), '');

  if (raw.isEmpty) {
    return const ParsedRoute(tab: null, id: null);
  }

  final parts = raw.split('/');
  final tabPart = parts[0];
  final idPart = parts.length > 1 ? parts[1] : null;
  final tab = isKnownTab(tabPart) ? tabPart : null;

  return ParsedRoute(
    tab: tab,
    id: (idPart != null && idPart.isNotEmpty) ? Uri.decodeComponent(idPart) : null,
  );
}

String formatRoute(String tab, String? id) =>
    id != null ? '#/$tab/${Uri.encodeComponent(id)}' : '#/$tab';
